PEP 544: Protocols

ilevkivskyi opened the pull request:
This adds static support for structural subtyping. Previous discussion is here https://github.com/python/typing/issues/11

Fixes #222

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+``Type[]`` with protocols
+-------------------------
+
+Variables and parameters annotated with ``Type[Proto]`` accept only concrete
```
How is `Type[C]` defined to work with a non-protocol ABCs? It seems that they should both work the same.

ilevkivskyi replied:
I think it could work like this for all ABCs. But it does not work like this in ``mypy `` now, and it is not specified in PEP 484. I feel like this should be specified here rather that in PEP 484.

By the way, there is an old high-priority issue for this https://github.com/python/mypy/issues/1843 and my PR https://github.com/python/mypy/pull/2853 that didn't receive a review for a month :-)

ambv replied:
I agree this PEP is the right place to finish specifying it. I find the existing explanation missing the "why". It is neatly demonstrated in https://github.com/python/mypy/issues/1843:
- we don't care for the additional *error* around passing a Protocol or an ABC
- we **do** care that treating `Type[P]` as any *concrete* subtype of `P` **silences** errors that signify invalid behavior only if ABCs or Protocols are passed

gvanrossum replied:
Hm, I'm not sure I agree that this should be specified in this PEP for cases other than protocols (though it's fine to discuss it in this thread). I'm also not sure about the sense of @ambv's second bullet, since "we care" is somehow ambiguous in my mind (do we find it a significant mistake or a significant feature?).

IIRC it's pretty subtle to prohibit passing an abstract class object (or a protocol object) to a parameter that requires something annotated with `Type[]`, since the thing you are passing might itself have such a type (e.g. `cls` in a class method).

ilevkivskyi replied:
@gvanrossum 
> Hm, I'm not sure I agree that this should be specified in this PEP for cases other than protocols.

As I understand, there are only two options:
* this behavior only applies to protocols
* it applies to all ABCs (including protocols)

I think the simplest solution is to say here than this behavior applies to protocols, but _don't_ 
say that it doesn't apply to ABCs. Then, when we decide on ABCs, we could update PEP 484 independently.

> IIRC it's pretty subtle to prohibit passing an abstract class object (or a protocol object) to a parameter that requires something annotated with ``Type[]``, since the thing you are passing might itself have such a type (e.g. ``cls`` in a class method).

My PR, that I mentioned above, actually takes care of this and few other edge cases, but it has never been reviewed...

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+
+Protocols are essentially anonymous. To emphasize this point, static type
+checkers might refuse protocol classes inside ``NewType()`` to avoid an
+illusion that a distinct type is provided::
```
This seems like an unnecessary restriction. `NewType` could define a non-protocol subtype of the protocol, so that compatibility only goes from the new type to the protocol but not the other way around. It would define a distinct type and would be consistent with how `NewType` works in general.

ilevkivskyi replied:
OK, I agree. The only limitation would be that ``NewType`` cannot provide a distinct protocol.

ddfisher replied:
I think this restriction makes sense, actually -- trying to construct a `NewType` from a protocol isn't cogent.  You probably want an alias instead.  Having `NewType` create a non-protocol subtype is surprising behavior and is *not* consistent with how `NewType` works in general, in my opinion.  We shouldn't support this without a clear need.

ambv replied:
I agree with @ddfisher, it doesn't fit my brain to have a NewType that is distinct from another Protocol, given that both are abstract and structural (and thus equivalent). To explain this differently:
- `What = NewType('What', SomeProtocol)` doesn't make sense because any concrete implementation of a protocol should both conform to SomeProtocol **and** What.

gvanrossum replied:
I'm with @ddfisher and @ambv here.

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+(``Iterable`` is statically equivalent to ``Iterable[Any]`).
+The metaclass of ``typing.Protocol`` will automatically add
+a ``__subclasshook__()`` that provides the same semantics
+for class and instance checks as for ``collections.abc`` classes::
```
I don't love the automatic `isinstance` check support, even though that's how the mypy `Protocol` class used to work. I don't think that they can be made to work reliably for arbitrary ABCs. Consider this example:

```py
class P(Protocol):
    @abstractmethod
    def common_method_name(self, x: int) -> int: ... 

class X:
    <a bunch of methods>
    def common_method_name(self) -> None: ...

def do_stuff(o: Union[P, X]) -> int:
    if isinstance(o, P):
        return o.common_method_name(1)  # oops what if it's an X instance instead?
    else:
        return (do something with an X)
```

Maybe a type checker can warn about this example, but more generally the `isinstance` check could be unreliable, except for things where there is a common signature convention such as `Iterable`.

Another potentially problematic case:

```py
class P(Protocol):
    x: int

class C:
    def initialize(self) -> None:
        self.x = 0

c1 = C()
c2 = C()
c2.initialize()
isinstance(c1, P)  # False
isinstance(c2, P)  # True

def f(x: Union[P, int]) -> None:
    if isinstance(x, P):
        # type of x is P here
        ...
    else:
        # type of x is int here?
        print(x + 1)

f(C())   # oops
```

Again, maybe we can live with this, but this could be pretty surprising. I'd argue that requiring an explicit class decorator would be better, since we can then attach warnings about problems like this in the documentation. The programmer would be able to evaluate whether the benefits outweigh the potential for confusion for each protocol and explicitly opt in -- but the default behavior would be safer.

ilevkivskyi replied:
> I'd argue that requiring an explicit class decorator would be better, since we can then attach warnings about problems like this in the documentation.

This is how I initially wrote this, since I intuitively felt that this could not be 100% reliable. Guido wanted to make it default, by now we have good examples to make this opt-in, so that I will revert this to class decorator.

ddfisher replied:
I'm not sure I understand how a class decorator would solve the problem.  Wouldn't that mean that some classes that implement the protocol (i.e. the decorated ones) will work properly with `isinstance` at runtime but others will not?

ilevkivskyi replied:
> classes that implement the protocol (i.e. the decorated ones)

The decorator is applied to protocol, not to implementation. This was in a previous draft. At runtime the decorator will add same ``__subclasshook__()`` to a protocol, that ``collections.abc`` classes have now. As Jukka pointed out, there are pros and cons of such approach, and a user could decide.

ddfisher replied:
How would the `__subclasshook__` work?

ilevkivskyi replied:
Basically, if a class has all necessary attributes, then it is a subclass. This is why I mention below that "type checkers will narrow types after such checks by the type __erased__ ``Proto`` (i.e. with all variables having type ``Any`` and all methods having type ``Callable[..., Any]``)" because we cannot reliably check the types of attributes at runtime.

This is how it is done in ``collections.abc`` see https://github.com/python/cpython/blob/master/Lib/_collections_abc.py#L72 and e.g. https://github.com/python/cpython/blob/master/Lib/_collections_abc.py#L93

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+          # Error! Can't use 'isinstance()' with subscripted protocols
+
+
+Using Protocols in Python 3.5 and Earlier
```
What about Python 2.7?

ilevkivskyi replied:
I think this could be used the same way for 2.7 - 3.5, modulo the fact that function type comments should be used in 2.7.

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+reasons:
+
+* Protocols don't have some properties of regular classes. In particular,
+  ``isinstance()`` is not always well-defined for protocols, whereas it is
```
A better argument would be something like this: `isinstance`, as defined for classes that don't explicitly extend `Protocol`, is based on the nominal hierarchy. In order to make everything a protocol by default *and* have `isinstance` work would require changing the semantics of `isinstance`, which won't happen.

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+  mainly because their interfaces are too large, complex or
+  implementation-oriented (for example, they may include de facto
+  private attributes and methods without a ``__`` prefix).
+* Most actually useful protocols in existing Python code seem to be implicit.
```
Additional reason: Many built-in functions only accept concrete instances of `int` (and subclass instances), and similarly for other built-in classes. Making `int` a structural type wouldn't be safe without major changes to the Python runtime, which won't happen.

JukkaL commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+References
+==========
+
+.. [typing]
```
Is it customary to include references to discussions that have happened about the PEP, such as in the typing issue tracker?

ilevkivskyi replied:
Yes, this is a good idea.

gvanrossum replied:
I wonder if we could do more deep linking to specific comments or subthreads? (No big deal if you think it's too much effort, but it might save some readers some time finding the right part of the long discussion in e.g. python/typing#11 -- including myself :-).

ilevkivskyi replied:
OK, will do this.

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+  finish(GoodJob()) # OK
+
+In addition, we propose to add another special type construct
+``All`` that represents intersection types. Although for normal types
```
I think we could include this for three reasons:
* I could easily imagine a function parameter that requires more than one protocol, especially if the preferred style is to have several simple protocols, rather than few complex. By the way, some classes in ``collections.abc`` are simple intersections, for example ``Collection == All[Sized, Iterable, Container]``.
* It could be easily implemented for protocols. I also propose to use ``All`` instead of ``Intersection`` to emphasize it only works for protocols (mnemonics: variable implements _all_ these protocols). On the contrary, I think that intersections of nominal classes would be _very_ rarely used.
* If we decide to include this later, then people might simply be not aware of this feature (this is a minor reason, but still we need to take this into account).

ddfisher replied:
I agree with Jukka -- I think it's confusing for `All` to work only with protocols, extending `All` to work with all types would add considerable complexity, and I'm not convinced there's a pressing need.

ambv replied:
I was pushing for `Intersection` for the original PEP because I saw the need for being able to intersect ABCs. Given that protocols themselves solve this, and we can use multiple inheritance as Jukka is suggesting, I agree that it's wiser to leave this out.

I would still *like* to have it, but that can be a separate PEP. I agree it's complex to implement for concrete types, and better yet, for combinations of both.

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+a compatible type signature.
+
+
+Protocol members
```
> This seems to imply that sequences no longer support iteration, since Sequence.__iter__ is not abstract and thus is not a protocol member. Is this intentional?

``Sequence`` _implements_ ``__iter__`` therefore it will be accepted wherever ``Iterable`` is expected. But asking for ``__iter__`` on ``Sequence`` indeed will fail.

This indeed looks problematic. Here is a solution that I could propose: we do not have such thing as non-protocol members. However, there will be two kinds of members, abstract and non-abstract. _All_ of them should be implemented in order to consider a class as structural (implicit) subtype.

The reason to have non-abstract methods is that one can get them "for free" using explicit subclassing provided abstract methods are implemented in subclass. So that for larger protocols like ``Sequence`` or ``Mapping`` one doesn't need to do
```python
def method(self):
    return super().method()
```
for every method.

This coincides with how ABCs work at runtime, so that this will be intuitive for someone who already worked with ABCs.

gvanrossum replied:
> Here is a solution that I could propose: we do not have such thing as non-protocol members.

That's now specified below in the PEP right?

ilevkivskyi replied:
> That's now specified below in the PEP right?

Yes.

ilevkivskyi commented:
@JukkaL Thank you for helpful comments! I agree with almost all things. I will make corresponding changes in the text soon. The only thing where we do not agree is ``All``, I still think we should have it.

Also we probably need to find a way how to indicate whether an abstract method has a default implementation or not _in stubs_.

ddfisher commented:
As a general comment, I don't think it's a good idea to standardize this before having an implementation. An implementation would allow us to understand what's difficult to implement, what sorts of unexpected interactions it has with other type-system features, what parts are confusing, and what parts are practically useful.  I think it will be very difficult to get an understanding of those things otherwise.

ddfisher commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+-----------------------------------
+
+To explicitly declare that a certain class implements the given protocols, they
+can be used as regular base classes. In this case:
```
I'm worried this could be pretty confusing: you could have a class that explicitly implements some protocols and implicitly implements others.  I don't know of any other language that allows this.

ddfisher commented on pep-0544.txt:
```diff
@@ -0,0 +1,949 @@
+  be provided at runtime.
+
+
+Changes in the typing module
```
How do you propose to handle the backwards-incompatibility problem for all currently released versions of Python?

ilevkivskyi replied:
The idea is that there should be no backward incompatibility. The classes in ``typing`` module (as well as in ``collections.abc``) already behave like protocols at runtime:
```python
class Whatever:
    def __iter__(self): return []
assert isinstance(Whatever(), Iterable)
```
as I mention below "Practically, few changes will be needed in ``typing`` since some of these classes already behave the necessary way at runtime", most of these will be changed only in typeshed.

Also, after I implement the changes proposed by Jukka, all code that passes mypy now, will still pass (unless I don't see some subtleties).

ilevkivskyi commented:
@ddfisher Of course I am not going to wait until this PEP is accepted :-)
I think implementation and discussion of specification should go in parallel. This is a big feature, but I hope I will make a first PR quite soon.

brettcannon commented:
@ilevkivskyi when you're ready for the first version to go in just let us know and we will commit it. Then you  can have discussions go to your personal fork and just send PRs for updates as needed.

ilevkivskyi commented:
@brettcannon I implemented most comments that appeared so far, but I would prefer to wait for input from @ambv

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+class would not turn the subclass into a protocol unless it also has
+``typing.Protocol`` as an explicit base class. Without this base, the class
+is "downgraded" to a regular ABC that cannot be used with structural
+subtyping. See section on `extending`_ for details of defining subprotocols.
```
This paragraph belongs to the next section (`_extending`).

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+
+    class Tree(Generic[T]):
+        def __init__(self, value: T,
+                           leafs: 'List[Tree[T]]') -> None:
```
s/leafs/leaves/

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+like trees in an abstract fashion::
+
+  class Traversable(Protocol):
+      leafs: Iterable['Traversable']
```
s/leafs/leaves/

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
+* Edge case: for recursive protocols, structural subtyping is decided
+  positively for situations where such decision depends on itself. Continuing
```
> for recursive protocols, structural subtyping is decided positively for situations where such decision depends on itself.

I see what you mean here but the wording is a little confusing.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+
+Assigning a protocol class to a variable is allowed if it is not explicitly
+typed, and such assignment creates a type alias. For non-protocol classes,
+the behavior of ``Type[]`` is not changed.
```
> For non-protocol classes, the behavior of ``Type[]`` is not changed.

No, we want to make ABCs behave the same way.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+  a protocol or not. Add a class attribute ``__protocol__ = True``
+  if that is the case. Verify that a protocol class only has protocol
+  base classes in the MRO (except for object).
+* Implement ``@runtime`` that adds all attributes to ``__subclsshook__()``.
```
`s/__subclsshook__/__subclasshook__/`

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+initialized in ways that are not visible to introspection
+(using ``setattr()``, assignment via ``self``, etc.). Still, some things like
+types of attributes will not be visible at runtime in Python 3.5 and earlier,
+but this looks like a reasonable limitation.
```
Yes.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+The protocols as described here are basically a minimal extension to
+the existing concept of ABCs. We argue that this is the way they should
+be understood, instead of as something that replaces Zope interfaces,
+for example. Attempting such interoperabilities will significantly
```
Zope interfaces are a *superset* of Protocols defined here, just using an incompatible syntax to define them. They necessarily use Attributes because pre-PEP 526 there was no other way to annotate this. Arguably in the 3.6+ world `zope.interface` could (optionally?) adopt the Protocol syntax.

In which case to achieve interoperability we'd only need to figure out a way to inform the type checker that "Interface" (or rather "InterfaceClass") is essentially "Protocol". This way we could type check successfully (pretty much ignoring `implements()` and treating those classes as *implicit* implementations. I think this is perfectly fine as the goal is **gradual typing**: if the type checker stays silent in some cases, that's OK. It just never should raise false positives.

Adapters, invariants, and runtime implements/provides checkers are obviously left out. This is a non-goal.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+
+  def f(x: Union[P, int]) -> None:
+      if isinstance(x, P):
+          # type of x is P here
```
> type of x is P here

or rather a concrete type that implements P ;-)

JukkaL replied:
The intention probably was that the _static type_ of `x` is `P`.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+------------------------------------------
+
+The problem with this is instance checks could be unreliable, except for
+situations where there is a common signature convention such as ``Iterable``.
```
Can you give an example of this unreliability? This sounds like a much stronger argument than the attribute assignment presented below (read comment about it).

ilevkivskyi replied:
I added a shorter version of Jukka's example here.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+  c = C()
+  isinstance(c1, P)  # False
+  c.initialize()
+  isinstance(c, P)  # True
```
This is not surprising to me and at runtime I can see this as being a useful feature. You can imagine:
```
class Client(Protocol):
    ...

class Connected(Protocol):
    connection: socket


...  # later in the code
isinstance(someClient, Connected)
```

Maybe we don't want to encourage this programming style but it definitely descibes some real world code, sigh.

JukkaL replied:
The problem here is that the runtime and static view of the type of `someClient` differ, potentially resulting in runtime type errors uncaught by type checkers, as shown by the code below this line. Since protocols are mostly about static type checking, I think it's reasonable to make restrictions that are biased towards letting static checkers do a better job.

ambv commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+
+  f(C())   # oops
+
+We argue that requiring an explicit class decorator would be better, since
```
That's the one part I'm not sure about. Having every Protocol behave consistently at runtime would make it easier to use.

JukkaL replied:
As discussed in the examples I gave, it would also make protocols more fragile and potentially surprising. My argument is that explicit here is better than implicit, even if it adds a bit of work for the programmer. Also, currently my working hypothesis is that protocols won't be very common in typical statically typed code -- they will be used every now and then, but not very frequently. If correct, the extra effort needed for adding the decorator would very little and basically insignificant. If this proves to be an incorrect hypothesis, it's much easier to add the missing implicit runtime check in the future than to remove an existing implicit runtime check, since the latter would break existing code. Basically, my reasoning is that if in doubt, leave it out from the first version until we have evidence that it would be useful.

gvanrossum replied:
OK, I am actually convinced by @JukkaL's argument here. (But we should probably apply a similar reasoning to cut out some other esoteric ideas, like variables.)

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+subclassed or registered. This is particularly difficult to do with library
+types as the type objects may be hidden deep in the implementation
+of the library. Moreover, extensive use of ABCs might impose additional
+runtime costs.
```
```python
class C(collections.abc.Iterable):
    def __iter__(self):
        return []
```
is 2x slower than
```python
class C:
    def __iter__(self):
        return []
```
The numbers for ``typing.Iterable`` are even worse, but I have some ideas on how to improve those. The 2x slowdown is still not much, so I use "extensive" and "might". If you think it is not necessary we could remove this altogether.

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,968 @@
+
+  Such behavior seems to be a perfect fit for both runtime and static behavior
+  of protocols. The main goal of this proposal is to support such behavior
+  statically. In addition, to allow users achieving such runtime behavior
```
http://ell.stackexchange.com/questions/11193/allow-to-infinitive-substantive-verb-ing says it probably should be "to achieve" here.

ilevkivskyi commented:
@brettcannon I think this could be merged now, unless there are some objections. I hope I addressed all points that were discussed. There may be some things that could be ironed out, but we can do this is additional PRs later.

gvanrossum commented:
Let's merge it. That does not mean I am accepting the PEP but is ready for
pyrhon-dev discussion.

--Guido (mobile)

On Mar 17, 2017 6:49 PM, "Ivan Levkivskyi" <notifications@github.com> wrote:

> @brettcannon <https://github.com/brettcannon> I think this could be
> merged now, unless there are some objections. I hope I addressed all points
> that were discussed. There may be some things that could be ironed out, but
> we can do this is additional PRs later.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/peps/pull/224#issuecomment-287508206>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACwrMpSk7CC9zp0bIc8O1Nyi3PafVb3kks5rmzhAgaJpZM4MaIqF>
> .
>

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+
+Note that although formally the implicit return type of a method with
+a trivial body is ``None``, type checker will not warn about above example,
+such convention is similar to how methods are defined in stub files.
```
Hmm, I don't really like the idea of "trivial bodies". Remember that when explicitly subclassing an ABC, abstract methods can still be called, through `super()`. An abstract method that shouldn't be called ought to raise `NotImplementedError`.

In stubs, mypy already doesn't type-check trivial bodies (though I don't recall its exact definition of "trivial" -- it probably doesn't exactly match what you have here). If we want to specify that behavior for all checkers reading stubs (which would seem to make sense) it should go in PEP 484, not here.

Or perhaps I am missing a use case for these that you've found in real code?

ilevkivskyi replied:
> An abstract method that shouldn't be called ought to raise ``NotImplementedError``.

OK, I think this is a reasonable requirement, I will update this part.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+annotations in the class body. Additional attributes *only* defined in
+the body of a method by assignment via ``self`` are not allowed. The rationale
+for this is that the protocol class implementation is often not shared by
+subtypes, so the interface should not depend on the default implementation.
```
Well in some sense that same rationale would apply to default (i.e. non-abstract) implementations.

A defensible position could be that as long as a type checker can verify that an implementation that doesn't explicitly inherit from the protocol class still defines the variable, there's no great reason to disallow variables, since they could just be interpreted as a shorthand for a setter and a getter method.

Also note that mypy (to take one example) doesn't actually check whether an attribute is always set -- it only checks that *if* it is used it has the right type.

All this gets pretty messy though and I am leaning towards not allowing protocol variables at all, other than read-only properties (possibly abstract). I don't recall having seen real-world examples of them (but if there are examples that might sway me).

ilevkivskyi replied:
Maybe this point needs some clarifications. First, I don't see why we need to force people to write:
```python
class Point2D(Protocol):
    @property
    def x(self) -> int:
        ...
    @property
    def y(self) -> int:
        ...
```
instead of just
```python
class Point2D(Protocol):
    x: int
    y: int
```
I think almost every function parameter annotated with a named tuple or a similar struct-like class could be replaced with a protocol annotation. (Also, this will be a natural counterpart of ``TypedDict``.) Second, I don't think we need to specify how "smart" type checkers should be. Maybe we could only mention that it is reasonable to expect that a type checker _will_ recognize
```python
class Coordinates:
    def __init__(self, x: int, y: int) -> None:
        self.x = x
        self.y = y
```
as implicitly implementing ``Point2D``.

ambv replied:
> All this gets pretty messy though and I am leaning towards not allowing protocol variables at all, other than read-only properties

Disallowing protocol variables entirely significantly limits the scope of structural typing. There's many examples in the stdlib alone (like the `stdout.buffer` field, named tuples and enums used structurally, etc.).

Only allowing read-only properties is less limiting but still so. More importantly, it is confusing to the user, like Ivan's example shows above:
* if we specify Point2D as a read-only property, should the type checker reject a writable implementation?
* if not, it feels strange to use a property to define something that is just an attribute in the implementation.

I can imagine users of this rewriting their attributes to be properties just to satisfy the protocol specification.

gvanrossum replied:
I'm not adamantly against protocol variables, but I think we should at least settle what to do about read-only attributes. When you use `@property` (not using the `@f.setter` syntax) or use PEP 526 syntax inside a`NamedTuple` you get a read-only attribute; otherwise PEP 526 syntax gets you a writable attribute. Which should be the default for Protocols, and how should you be able to declare the other kind?

Note that this is *not* about immutability -- that's in the programmer's head (or perhaps guided by whether `__hash__` is defined). This is purely about whether attribute assignment should be allowed.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+Explicitly declaring implementation
+-----------------------------------
+
+To explicitly declare that a certain class implements the given protocols,
```
I'd use a singular protocol here and below.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+
+Abstract methods with trivial bodies are recognized by type checkers as
+having no default implementation and can't be used via ``super()`` in
+explicit subclasses. The default implementations can not be used if
```
can not -> cannot

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+    represent(nice) # OK
+    represent(another) # Also OK
+
+Note that there is no conceptual difference between explicit and implicit
```
Well, hm, there seem to be plenty of subtle differences, e.g. the requirement to implement everything only in an implicit subclass.

ilevkivskyi replied:
> e.g. the requirement to implement everything only in an implicit subclass

This is exactly what I mean on the next line.

gvanrossum replied:
Still, starting with "there is no conceptual difference..." seems too strong.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+classes. In this case methods are resolved using normal MRO and a type checker
+verifies that all subtyping are correct. The semantics of ``@abstractmethod``
+is not changed, all of them must be implemented by an explicit subclass
+before it could be instantiated.
```
could -> can

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+-------------------------------
+
+The general philosophy is that protocols are mostly like regular ABCs,
+but a static type checker will handle them specially. Subclassing a protocol
```
(At runtime maybe we could say that they behave like an implicit `register()` call exists?)

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+class would not turn the subclass into a protocol unless it also has
+``typing.Protocol`` as an explicit base class. Without this base, the class
+is "downgraded" to a regular ABC that cannot be used with structural
+subtyping.
```
I think we need a rationale for this rule. I think there probably is one, but it needs to be stated clearly, and I'm not sure what it is -- perhaps that we don't want to accidentally have some class act as a protocol just because one of its base classes happens to be one. That is, we still (slightly) prefer nominal subtyping over structural subtyping (in the static typing world).

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+
+Now the protocol ``SizedAndCloseable`` is a protocol with two methods,
+``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
+this would be a regular (non-protocol) class that must implement ``Sized``.
```
Hm, this example makes this requirement feel like a (slight) nuisance -- in this particular example it's pretty clear that it's meant to define a protocol?

ilevkivskyi replied:
It may be clear for a human (taking into account the name of subclass). But for a type checker there are two equally possible options: a user actually wanted a protocol, or just forgot to implement a protocol method. As you said above:
> we don't want to accidentally have some class act as a protocol just because one of its base classes happens to be one

I think explicit is better than implicit applies here.

gvanrossum replied:
Fair enough. Objection withdrawn.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
+this would be a regular (non-protocol) class that must implement ``Sized``.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class.
```
This seems to make sense, but does it really? There's no such rule for ABCs. What harm would there be in a protocol inheriting from a regular class? Hm, I guess there's ambiguity over whether the methods in that regular class all become protocol methods with default implementations. Hm... I guess it does make sense, but maybe an explicit example showing the rationale would help.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+
+  from typing import Sized
+
+  class SupportsClose(...): ...  # Like above
```
I would just write this out rather than letting the reader guess to where exactly this is referring.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+
+Recursive protocols are also supported. Forward references to the protocol
+class names can be given as strings as specified by PEP 484. Recursive
+protocols will be useful for representing self-referential data structures
```
will be -> are

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+Subtyping relationships with other types
+----------------------------------------
+
+Protocols cannot be instantiated, so there are no values with
```
That sounds a little weird -- certainly an instance of a class that implements a protocol is a value of that protocol type? Maybe all you need is to add "exactly" somewhere to the sentence.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+      def __iter__(self) -> Iterator[T]:
+          ...
+
+Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
```
I think this deserves calling out, rather than just "Note".

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+    def walk(graph: Traversable) -> None:
+        ...
+    tree: Tree[float] = Tree(0, [])
+    walk(tree) # OK, 'Tree[float]' is a subtype of 'Traversable'
```
Hm, I'm not sure I like this example. Isn't the problem more that `Traversible` ought to be generic but isn't in this example?

ilevkivskyi replied:
> Isn't the problem more that ``Traversible`` ought to be generic but isn't in this example?

No, this "problem" (I would rather say "rule") applies to all recursive protocols. For example:
```python
class Linked(Protocol):
    child: 'Linked'

class C:
    child: 'C'

x: Linked
x = C() # This should be OK, according to the "rule"
```
I think we just need a simpler example here.

gvanrossum replied:
Agreed. The generics just distract. (Maybe we need a simple example first and then one showing how to combine this with generics?)

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+  class Exitable(Protocol):
+      def exit(self) -> int:
+          ...
+  class Quitable(Protocol):
```
I'd write "Quittable" -- just like it's "quitting" vs. "exiting". (English is never regular. :-)

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+
+  def finish(task: Union[Exitable, Quitable]) -> int:
+      ...
+  class GoodJob:
```
Can you find a different name for this? "Good job" is such an incredibly boring cliché...

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+  cached_func((1, 2, 3)) # OK, tuple is both hashable and sequence
+
+If this will prove to be a widely used scenario, then a special
+intersection type construct may be added in future as specified by PEP 483,
```
may -> could

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+
+  UserId = NewType('UserId', Id)  # Error, can't provide distinct type
+
+On the contrary, type aliases are fully supported, including generic type
```
On the contrary -> In contrast

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+  from typing import runtime, Protocol
+
+  @runtime
+  class Closeable(Protocol):
```
(Oh, I think I'd prefer Closable -- no 'e'.)

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+      if isinstance(items, Iterator):
+          # 'items' have type 'Iterator[int]' here
+      elif isinstance(items, Sequence[int]):
+          # Error! Can't use 'isinstance()' with subscripted protocols
```
I believe that's not unique to protocols -- it's the same for other generic classes.

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+      def d(self) -> int:   # ... or it can be abstract
+          return 0
+
+In Python 2.7 the function type comments should be used as per PEP 484.
```
I think they should also be allowed for Python 3 (even 3.6+) and the text earlier that introduces these should mention this. (It's fine of course to recommend PEP 526 syntax and to use it in all examples except one showing the fallback syntax.)

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+Use assignments to check explicitly that a class implements a protocol
+----------------------------------------------------------------------
+
+In Go language the explicit checks for implementation are performed
```
"In Go language" -> either "In Go" or "In the Go language"

gvanrossum commented on pep-0544.txt:
```diff
@@ -0,0 +1,1006 @@
+
+  def do_stuff(o: Union[P, X]) -> int:
+      if isinstance(o, P):
+          return o.common_method_name(1)  # oops, what if it's an X instance?
```
So this more or less assumes that the runtime isinstance check doesn't compare the arguments. But what if it did?

Or are you saying that in this case if `P` was decorated with `@runtime` the isinstance check would still be broken?

ilevkivskyi replied:
> Or are you saying that in this case if ``P`` was decorated with ``@runtime`` the isinstance check would still be broken?

I think yes, for example now ``isinstance(obj, Iterable)`` returns ``True`` for a class with ``__iter__ = 42``. And we can't make ``@runtime`` much better because not all information is available statically. As Jukka said, it is not a big problem since static type checker will warn about this, but big enough to make support for ``isinstance()`` with protocols opt-in.

gvanrossum replied:
OK, I'm convinced (also this review UI is very confusing, it seems we are having discussion on different subcommits of the same PR).

I still don't like the name `@runtime` though -- it's too generic. Can we brainstorm about a better name?

gvanrossum commented:
There's a couple more things that I should bring up.

It looks like the mypy core team (all Dropboxers) is going to be very busy with other features, to the point where we probably won't have time to even review an implementation of this PEP for the foreseeable future. Now, this is all subject to various priorities that may still change, and perhaps the implementation is super simple, but if it's anything like I expect, even a complete, finished implementation (and one that allows importing `Protocol` from `mypy_extensions` so it can be made to work with the stdlib versions of typing in earlier Python 3.5 and 3.6 releases) would still take up quite some time to review, time which we just may not have.

There's also the absolute requirement to send this PEP to python-dev for discussion. Since I'm not an author we don't need a BDFL-delegate -- we just need to be careful to listen to feedback. And while I am aware that my own time is a *particularly* scarce resource, I still don't want to delegate this particular topic.

ilevkivskyi commented:
@gvanrossum 
> There's also the absolute requirement to send this PEP to python-dev for discussion.

OK, I will do this now.

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+of a protocol, it is said to implement the protocol and to be compatible
+with a protocol. If a class is compatible with a protocol but the protocol
+is not included in the MRO, the class is an *implicit* subtype
+of the protocol.
```
Yes. I will add a note that type checkers should understand setting attributes to ``None`` in the sense of [data model docs](https://docs.python.org/3/reference/datamodel.html#special-method-names): "Setting a special method to ``None`` indicates that the corresponding operation is not available."

ilevkivskyi commented on pep-0544.txt:
```diff
@@ -0,0 +1,967 @@
+* A protocol is never a subtype of a concrete type.
+* A concrete type or a protocol ``X`` is a subtype of another protocol ``P``
+  if and only if ``X`` implements all protocol members of ``P``. In other
+  words, subtyping with respect to a protocol is always structural.
```
> Maybe it would be less ambiguous if you separated the two cases into separate bullets.

I agree.

ilevkivskyi commented:
@gvanrossum 
> we probably won't have time to even review an implementation of this PEP for the foreseeable future.

This sounds a bit frustrating, but what do you mean by "foreseeable future", few years or few days?
Also, @ddfisher said above in https://github.com/python/peps/pull/224#issuecomment-286907763 that "I don't think it's a good idea to standardize this before having an implementation", so that I thought that I should finish an implementation soon (even if not 100% complete).
I am a bit lost here.

gvanrossum commented:
> This sounds a bit frustrating, but what do you mean by "foreseeable future", few years or few days?

A few months at least, is my best estimate based on discussions we had in our project planning (not all [arts of the project are open source).

I agree it's frustrating -- it is for us too, but we need to acknowledge that we have limited velocity.

> Also, @ddfisher said above in #224 (comment) that "I don't think it's a good idea to standardize this before having an implementation", so that I thought that I should finish an implementation soon (even if not 100% complete).

I believe David is more skeptical of this PEP than Jukka and I. But even so, the extra frustrating thing is that, indeed, the PEP process (by convention at least) prefers to have a reference implementation to support a proposal, without any guarantees that the reference implementation will ever be deployed. This is similar to the RFC-based IETF standardization process for internet protocols -- and the similarity is not coincidental, we modeled Python's PEP process after the RFC process (with which several core Python devs were highly familiar between 1995-2000).

> I am a bit lost here.

I'm truly sorry. It is frustrating for all of us (as are my continued health issues that reduce my productivity dramatically).

ambv commented:
@gvanrossum

> we need to acknowledge that we have limited velocity.

What would it take for the project to nominate a few non-Dropbox core contributors?

> A few months at least

This is not a huge obstacle for inclusion in Python 3.7, as long as we get to it sometime this year. Also, a reference implementation for the purpose of this PEP can be on a "branch" of mypy. In other words, for the PEP reviewers to consider it, it doesn't have to be merged. This was pretty common in the past.

@ilevkivskyi, I will provide preliminary review and use it internally before it gets merged (we are living on master anyway).

ilevkivskyi commented:
OK, what about this way: I will have a branch in my fork of mypy, where I will build a reference implementation. People will be able to play with it (I will take care to merge mypy master frequently, so that the branch will be fully functional). And it is fine with me if it will be not merged for few months. To avoid a potential huge (and impossible to review) PR if/when the PEP is accepted, I will cherry-pick few smaller branches and make few sequential PRs.

@gvanrossum @ambv @JukkaL does this sound reasonable?

gvanrossum commented:
It sounds reasonable (though I cannot make any hard promises).

CC @gnprice.
